crates/turborepo-lib/src/task_hash.rs

```diff
@@ -0,0 +1,318 @@
+    package_task_hashes: HashMap<TaskId<'static>, String>,
+    package_task_framework: HashMap<TaskId<'static>, String>,
+    package_task_outputs: HashMap<TaskId<'static>, Vec<AnchoredSystemPathBuf>>,
+}
+
+/// Caches package-inputs hashes, and package-task hashes.
+pub struct TaskHasher<'a> {
+    opts: &'a Opts<'a>,
+    env_at_execution_start: &'a EnvironmentVariableMap,
+    global_hash: &'a str,
+
+    task_hash_tracker: Mutex<TaskHashTracker>,
+}
+
+impl<'a> TaskHasher<'a> {
+    pub fn new(
+        opts: &'a Opts,
+        env_at_execution_start: &'a EnvironmentVariableMap,
+        global_hash: &'a str,
+    ) -> Self {
+        Self {
+            opts,
+            env_at_execution_start,
+            global_hash,
+            task_hash_tracker: Mutex::new(TaskHashTracker::default()),
+        }
+    }
+
+    /// Calculates the package inputs hash, i.e. the hash of
+    /// the files for a given task.
+    /// This is a little bit of an expensive operation,
+    /// so in the future we should look into caching it.
+    pub async fn calculate_package_inputs_hash(
+        &self,
+        scm: &SCM,
+        task_id: &TaskId<'static>,
+        task_definition: &TaskDefinition,
+        pkg: &WorkspaceInfo,
+        repo_root: &AbsoluteSystemPath,
+    ) -> Result<Option<String>, Error> {
+        if task_id.package() == ROOT_PKG_NAME {
+            return Ok(None);
+        }
+
+        let workspace_name = WorkspaceName::Other(task_id.package().to_string());
+
+        let package_file_hash_inputs = PackageFileHashInputs {
+            task_id: task_id.clone(),
+            task_definition,
+            workspace_name,
+        };
+
+        let package_path = pkg
+            .package_json_path
+            .parent()
+            .unwrap_or_else(|| AnchoredSystemPath::new("").unwrap());
+
+        let mut hash_object = scm.get_package_file_hashes(
+            &repo_root,
+            package_path,
+            &package_file_hash_inputs.task_definition.inputs,
+        )?;
+
+        if !package_file_hash_inputs.task_definition.dot_env.is_empty() {
+            let package_path = pkg
+                .package_json_path
+                .parent()
+                .unwrap_or_else(|| AnchoredSystemPath::new("").unwrap());
+            let absolute_package_path = repo_root.resolve(package_path);
+            let dot_env_object = scm.hash_existing_of(
+                &absolute_package_path,
+                package_file_hash_inputs
+                    .task_definition
+                    .dot_env
+                    .iter()
+                    .map(|p| p.to_anchored_system_path_buf()),
+            )?;
+
+            for (key, value) in dot_env_object {
+                hash_object.insert(key, value);
+            }
+        }
+
+        let file_hashes = FileHashes(hash_object);
+        let hash = file_hashes.clone().hash();
+
+        Ok(Some(hash))
+    }
+
+    pub async fn calculate_task_hash(
```

Comment:
This can also be made sync if the other changes suggested are made.
```suggestion
    pub fn calculate_task_hash(
```